packages/indy-sdk-to-askar-migration/src/IndySdkToAskarMigrationUpdater.ts

```diff
@@ -0,0 +1,363 @@
+    }
+
+    /**
+     * Since we used `copy` to get the file, it should still be there. We
+     * double-check here to be sure.
+     */
+    if (!(await this.fs.exists(this.dbPath))) {
+      return
+    } else {
+      this.agent.config.logger?.trace(`Moving '${this.backupFile}' back to the original path: '${this.dbPath}`)
+
+      // Move the backedup file back to the original path
+      await this.fs.copyFile(this.backupFile, this.dbPath)
+
+      this.agent.config.logger?.trace(`Cleaned up the backed up file at '${this.backupFile}'`)
+    }
+  }
+
+  // Delete the backup as `this.fs.copyFile` only copies and no deletion
+  // Since we use `tempPath` which is cleared when certain events happen,
+  // e.g. cron-job and system restart (depending on the os) we could omit
+  // this call `await this.fs.delete(this.backupFile)`.
+  private async cleanBackup() {
+    this.agent.config.logger?.trace(`Deleting the backup file at '${this.backupFile}'`)
+    await this.fs.delete(this.backupFile)
+  }
+
+  /**
+   * Move the migrated and updated database file to the new location according
+   * to the `FileSystem.dataPath`.
+   */
+  private async moveToNewLocation() {
+    const src = this.dbPath
+    // New path for the database
+    const dest = this.newWalletPath
+
+    // create the wallet directory
+    await this.fs.createDirectory(dest)
+
+    this.agent.config.logger?.trace(`Moving upgraded database from ${src} to ${dest}`)
+
+    // Copy the file from the database path to the new location
+    await this.fs.copyFile(src, dest)
+
+    // Delete the original, only if specified by the user
+    if (this.deleteOnFinish) await this.fs.delete(this.dbPath)
+  }
+
+  /**
+   * Function that updates the values from an indy-sdk structure to the new askar stucture.
+   *
+   * - Assert that the paths that will be used are free
+   * - Create a backup of the database
+   * - Update the Keys
+   * - Update the Master Secret (Link Sercet)
+   * - Update the credentials
+   * If any of those failed:
+   *   - Revert the database
+   * - Clear the backup from the temporary directory
+   */
+  public async update() {
+    await this.assertDestinationsAreFree()
+
+    await this.backupDatabase()
+    try {
+      await this.updateKeys()
+      await this.updateMasterSecret()
+      await this.updateCredentials()
+
+      await this.moveToNewLocation()
+    } catch (cause) {
+      this.agent.config.logger?.error('Migration failed. Reverting state.')
+
+      await this.revertDatabase()
+
+      throw new IndySdkToAskarMigrationError('Migration failed. State has been reverted.', { cause })
+    } finally {
+      await this.cleanBackup()
+    }
+  }
+
+  private async updateKeys() {
+    const category = 'Indy::Key'
+
+    this.agent.config.logger?.trace(`[indy-sdk-to-askar-migration]: Updating category: ${category}`)
+
+    let updateCount = 0
+    const session = this.store.transaction()
+    for (;;) {
```

Comment:
Nice to see this in TypeScript!

Reply:
Yeah, apparently it was a fix for aslant not allowing `while(true)` :). This might be some code-smell, but for now I just wanted to be consistent with aca-py. Also, never knew this was also possible in TS.